src/main/scala/cognite/spark/v1/SdkV1Rdd.scala

```diff
@@ -81,33 +71,27 @@ case class SdkV1Rdd[A, I](
     }(drainContext)
 
     queueIterator(queue, streamsToQueue) {
-      if (!threadPool.isShutdown) {
-        threadPool.shutdown()
-      }
       if (!drainPool.isShutdown) {
         drainPool.shutdown()
       }
     }
   }
 
-  // To avoid draining all streams from CDF completely and then building the Iterator, we read the
-  // streams in chunks and add the chunks to a queue continuously.
+  // We avoid draining all streams from CDF completely and then building the Iterator,
+  // by using a blocking EitherQueue.
   def enqueueStreamResults(
       stream: Stream[IO, A],
       queue: EitherQueue,
-      processedIds: mutable.Set[I],
-      singleThreadedCs: ContextShift[IO]): Stream[IO, Unit] =
-    for {
-      // Ensure that only one chunk is processed at a time
-      s <- stream.chunks.flatMap(c => Stream.eval(singleThreadedCs.shift *> IO.pure(c)))
-      put <- {
-        // Filter out and keep track of already seen items
-        // since overlapping pushdown filters may read duplicates.
-        val freshIds = s.toVector.filterNot(i => processedIds.contains(uniqueId(i)))
-        processedIds ++= freshIds.map(i => uniqueId(i))
-        Stream.eval(IO(queue.put(Right(freshIds))))
+      processedIds: ConcurrentHashMap[I, Unit]): Stream[IO, Unit] =
+    stream.chunks.parEvalMapUnordered(config.parallelismPerPartition * 2) { chunk =>
+      val freshIds = chunk.toVector.filterNot(i => processedIds.containsKey(uniqueId(i)))
```

Comment:
I know we did a `toVector` before, however I'm unsure if that's needed, and only ends up being an unnecessary copy operation (though only a 10k * IntPtr.size).

Reply:
Good point, it's easy to work with Chunks directly.